rust/connlib/tunnel/src/client.rs

```diff
@@ -1185,43 +1284,74 @@ impl ClientState {
-                                self.tcp_dns_server.send_message(
-                                    query.socket,
-                                    dns::servfail(message.for_slice_ref())
-                                ),
-                                "Failed to send TCP DNS response: {}"
-                            );
-                            return;
-                        }
-                    };
-
-                    let existing = self
-                        .tcp_dns_sockets_by_upstream_and_query_id
-                        .insert((server, query_id), query.socket);
-
-                    debug_assert!(existing.is_none(), "Query IDs should be unique");
+                    self.forward_tcp_dns_query_to_new_upstream_via_tunnel(server, query);
 
                     return;
                 }
 
                 tracing::trace!(%server, %query_id, "Forwarding TCP DNS query");
 
                 self.buffered_dns_queries
-                    .push_back(dns::RecursiveQuery::via_tcp(query.socket, server, message));
+                    .push_back(dns::RecursiveQuery::via_tcp(
+                        query.socket,
+                        server,
+                        query.message,
+                    ));
+            }
+            dns::ResolveStrategy::RecurseSite(resource) => {
+                let Some(gateway) =
+                    peer_by_resource_mut(&self.resources_gateways, &mut self.peers, resource)
+                else {
+                    self.on_not_connected_resource(
+                        resource,
+                        ConnectionTrigger::TcpDnsQueryForSite(query),
+                        now,
+                    );
+                    return;
+                };
+
+                let server = gateway.tun_dns_server_endpoint(query.local.ip());
+
+                self.forward_tcp_dns_query_to_new_upstream_via_tunnel(server, query);
             }
         };
     }
 
+    fn forward_tcp_dns_query_to_new_upstream_via_tunnel(
+        &mut self,
+        server: SocketAddr,
+        query: dns_over_tcp::Query,
+    ) {
+        let query_id = query.message.header().id();
+
+        match self
+            .tcp_dns_client
+            .send_query(server, query.message.clone())
+        {
+            Ok(()) => {}
+            Err(e) => {
+                tracing::warn!(
+                    "Failed to send recursive TCP DNS query to upstream resolver: {e:#}"
+                );
+
+                unwrap_or_debug!(
+                    self.tcp_dns_server
+                        .send_message(query.socket, dns::servfail(query.message.for_slice_ref())),
+                    "Failed to send TCP DNS response: {}"
+                );
+                return;
+            }
+        };
+
+        let existing = self
+            .tcp_dns_sockets_by_upstream_and_query_id
+            .insert((server, query_id), query.socket);
+
+        debug_assert!(existing.is_none(), "Query IDs should be unique");
```

Comment:
I remember @conectado running into a nasty bug on Windows where DNS query IDs were notably **not** unique and (I believe) we had to also identify them by the name being queried or something.

Just flagging this in case it might be relevant here.

Reply:
These are scoped by upstream server too. If the same client is reusing IDs to the same server, there is no way we can correctly associate the response.